source/extensions/path/pattern_template_lib/pattern_template_internal.h

```diff
@@ -0,0 +1,147 @@
+#pragma once
+
+#include <optional>
+#include <string>
+#include <variant>
+#include <vector>
+
+#include "absl/container/flat_hash_set.h"
+#include "absl/status/status.h"
+#include "absl/status/statusor.h"
+#include "absl/strings/string_view.h"
+#include "absl/types/optional.h"
+#include "absl/types/variant.h"
+#include "re2/re2.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace PatternTemplate {
+
+namespace Internal {
+
+using Literal = absl::string_view;
+
+/**
+ * Determines what operations to use on the input pattern segment
+ */
+enum class Operator { KPathGlob, KTextGlob };
+
+/**
+ * Represents a segment of the rewritten URL, including any path segments,
+ * slash and prefix.
+ */
+struct RewriteSegment {
+
+  // A string that will be concatenated for rewrite.
+  absl::string_view literal_;
+
+  // Represents an index into the RE2 capture which value should be used
+  // to construct the rewritten URL. Note that the index should be greater
+  // than 0 as 0 index into the whole match RE2 pattern.
+  int capture_index_;
+};
+
+/**
+ * Represents a pattern variable. Variables are included in both path match and rewrite paths.
+ */
+struct Variable {
+  absl::string_view var_name_;
+  std::vector<absl::variant<Operator, Literal>> var_match_;
+
+  Variable(absl::string_view name, std::vector<absl::variant<Operator, Literal>> match)
+      : var_name_(name), var_match_(match) {}
+
+  std::string debugString() const;
+};
+
+using ParsedSegment = absl::variant<Operator, Variable, Literal>;
+
+/**
+ * Represents the parsed path including literals and variables.
+ */
+struct ParsedUrlPattern {
+  std::vector<ParsedSegment> parsed_segments_;
+  absl::optional<absl::string_view> suffix_;
+  absl::flat_hash_set<absl::string_view> captured_variables_;
+
+  std::string debugString() const;
+};
+
+/**
+ * Check if literal is valid
```

Comment:
What would make a literal invalid? (looking at the .cc file, I'm not sure). Do we have unit test coverage for this method?